File: Nowcoder/Multi-university2020/round8/K.cpp
#include <algorithm>
#include <cctype>
#include <cmath>
#include <cstdio>
#include <cstring>
#include <iostream>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <vector>
#define mst(a, b) memset((a), (b), sizeof(a))
#define debug printf("debug\n")
#define INF 0x3f3f3f3f
#define lson lef, mid, rt << 1
#define rson mid + 1, rig, rt << 1 | 1
const int maxn = 1e5 + 5;
using namespace std;
typedef __int128 ll;
typedef unsigned long long ull;
ll a[maxn], b[maxn];
ll read()
{
    char ch;
    ll s = 1;
    do {
        ch = getchar();
        if (ch == '-')
            s = -1;
    } while (ch > '9' || ch < '0');
    ll ret = 0;
    while (ch >= '0' && ch <= '9') {
        ret = ret * 10 + ch - '0';
        ch = getchar();
    }
    return ret * s;
}
void write(ll x)
{
    if (x < 0) {
        putchar('-');
        x = -x;
    }
    if (x > 9)
        write(x / 10);
    putchar(x % 10 + '0');
}
void writeLn(ll x)
{
    write(x), puts("");
}
ll pre[maxn];
struct node {
    ll val;
    int id;
    bool operator<(const node& nn) const
    {
        return val < nn.val;
    }
};
priority_queue<node> pq;
int main()
{
    int kase = 0;
    int T;
    scanf("%d", &T);
    while (T--) {
        int n;
        scanf("%d", &n);
        pre[0] = 0;
        while (!pq.empty())
            pq.pop();
        for (int i = 1; i <= n; i++) {
            a[i] = read();
            pre[i] = pre[i - 1] + a[i];
            pq.push({ pre[i], i });
        }
        ll ges;
        b[1] = read(), ges = b[1];
        for (int i = 2; i <= n; i++) {
            b[i] = read();
            b[i] = min(b[i], b[i - 1]);
        }
        ll ans = 0;
        ll base = 0;
        while (!pq.empty()) {
            node now = pq.top();
            pq.pop();
            int idx = now.id;
            if (base < b[idx]) {
                ans += now.val * (b[idx] - base);
                base += (b[idx] - base);
            }
        }
        printf("Case #%d: ", ++kase), write(ges), printf(" "), write(ans), puts("");
    }
    return 0;
}

File: Nowcoder/Multi-university2020/round8/K.py
class Node:
    def __init__(self, b, p, a):
        self.b = b
        self.p = p
        self.a = a

ns = [Node(0,0,0)]
stk = []
v = []

T = int(input())
kase = 0
while T > 0:
    T = T - 1
    n = int(input())
    for i in range(1,n+1):
        a = int(input())
        tmpa = ns[-1].a
        ns.append(Node(0,0,tmpa+a))
    for i in range(1,n+1):
        ns[i].b = int(input())
        ns[i].p = i
        if(i!=1):
            ns[i].b = min(ns[i-1].b,ns[i].b)
    while len(stk) != 0:
        stk.pop()
    v.clear()
    for i in reversed(range(1,n+1)):
        while len(stk) != 0:
            l = stk[0]
            if ns[i].a >= ns[l].a:
                stk.pop()
            else:
                break
        stk.append(i)

    while len(stk) != 0:
        v.append(stk[-1])
        stk.pop()
    length = len(v)
    ans = 0
    for i in reversed(range(length)):
        if i == length - 1:
            ans = ans + ns[v[i]].a * ns[v[i]].b
        else:
            ans = ans + ns[v[i]].a * (ns[v[i]].b - ns[v[i+1]].b)
    print('Case #{}: {} {}'.format(kase, ns[0].b,ans))
    kase = kase + 1
    
